File: app/star_pass_cli/_confirm.py
#!/usr/bin/env python3
""" The gate in front of the things that cannot be undone.

    A send writes into a live volunteer system and Amplify has no way
    to take a shift back, so somebody reads what is about to happen and
    says yes.  The confirmation restates the count, the window and the
    opportunities on the line above the question rather than asking a
    bare "are you sure".

    A deletion is the second.  It destroys nothing Amplify holds -- a
    run that sent anything is refused one -- but what it destroys here
    is gone, so it is put the same way and restates the run it is
    about.

    The answer is yes or no, never a typed number: typing a number
    tests typing rather than attention.

    No answer is not yes.  Where there is no terminal to answer from --
    a script, a scheduled job, a pipe -- the send is refused rather
    than carried out, and no flag turns that into a yes.
"""

# Imports - Python Standard Library
import sys

# Imports - Local
from ._output import write

# What counts as yes.  Everything else, including an empty line, is
# no: the safe answer is the one somebody gets by pressing return
# without reading, and by pressing anything at all without meaning it.
AFFIRMATIVE = frozenset({'y', 'yes'})

# What the question ends with, so the safe answer is visibly the
# default one.
CHOICES = '[y/N]'

# What a caller is told when nothing can answer.  One for each thing
# asked about, because a caller told the wrong reason would go looking
# for a send that is not there.
NO_TERMINAL_TO_SEND = (
    'This command writes to Amplify and cannot be undone, so it asks '
    'first -- and there is no terminal here to ask. Run it where '
    'somebody can answer.'
)
NO_TERMINAL_TO_DELETE = (
    'This command deletes a run and cannot be undone, so it asks '
    'first -- and there is no terminal here to ask. Run it where '
    'somebody can answer.'
)


class ConfirmationUnavailable(Exception):
    """ There is nobody to ask, so the answer cannot be assumed.

        Raised rather than answered as a no, because the two are
        different things: a person who said no made a decision, and
        this is a command that could not put the question.
    """


def confirmed(
        question: str,
        unavailable: str
) -> bool:
    """ Put a question and return whether the answer was yes.

        Args:
            question (str):
                What to ask, without the choices, which are added.

            unavailable (str):
                What to say when there is nobody to ask, which names
                what was being asked about rather than leaving a
                caller to guess.

        Raises:
            ConfirmationUnavailable:
                If there is no terminal to answer from.

        Returns:
            answered (bool):
                Whether the answer was one of 'AFFIRMATIVE'.
    """

    if not sys.stdin.isatty():
        raise ConfirmationUnavailable(unavailable)

    write(f'{question} {CHOICES} ', end='')

    return input().strip().lower() in AFFIRMATIVE

File: web/js/runs/confirm.js
/* What a deletion restates before it happens.
 *
 * The second thing that cannot be undone, and it is put the same way
 * as the first, through the same dialog: what is about to go is on
 * the card above the button, so somebody reads it rather than
 * answering a bare "are you sure".  The counts are the point -- a run
 * holding nothing and a run holding a reviewed month read the same
 * way in an identifier alone.
 *
 * **What this destroys is not what Amplify holds.**  A run that sent
 * anything is refused a deletion outright and is offered no control
 * at all, so nothing reachable from here can take a shift out of a
 * volunteer's calendar.  What it destroys is the record on this side,
 * and that is gone.  The wording says so in both directions: what
 * goes with the run, and what stays behind because it outlives it.
 */

import { ConfirmDialog } from '../confirm.js';
import { counted, windowText } from '../format.js';
import { el } from '../dom.js';
import { runStatusTag } from './summary.js';

/* The question, which is the command line's, word for word: one
 * deletion asked about two ways would be two decisions to keep
 * agreeing with each other. */
const QUESTION = 'Delete this run and everything in it?';

/* Said last, above the buttons, in the alert colour. */
const NO_UNDO = 'This cannot be undone.';

/* What goes, and what does not. The titles a window did not match
 * belong to no run: what the data model is missing outlives the
 * window that revealed it. */
const WHAT_GOES = (
  'Its revisions, events, opportunities, change log and jobs go with '
  + 'it. The titles its window did not match stay, because what the '
  + 'shift data model is missing outlives the run that found it.'
);

/* What the button that goes ahead says. Never a bare "Delete": it
 * names what it is about to do, the way the send's does. */
const CONFIRM = 'Delete this run';

/** Return the sentence restating what is about to go.
 *
 * @param {Object} run The run, as the contract lists it.
 * @returns {string} The line.
 */
export function restatement(run) {
  const { counts } = run;

  return (
    `${counted(counts.events, 'event')}, `
    + `${counted(counts.shifts, 'shift')}, `
    + `${counted(run.currentRevision, 'revision')}.`
  );
}

/** Return the confirmation in front of deleting a run.
 *
 * @param {Object} what What is being confirmed.
 * @param {Object} what.run The run about to go.
 * @param {Object} handlers What its two buttons do.
 * @param {Function} handlers.onConfirm Go ahead.
 * @param {Function} handlers.onCancel Close and change nothing.
 * @returns {ConfirmDialog} The dialog, ready to open.
 */
export function deleteDialog({ run }, handlers) {
  return new ConfirmDialog(
    {
      title: QUESTION,
      about: `${run.calendar} · ${windowText(run.window)}`,
      body: [
        el(
          'div',
          { class: 'delete-holds' },
          el('span', { class: 'confirm-line', text: restatement(run) }),
          runStatusTag(run)
        ),
        el('p', { class: 'muted meta', text: WHAT_GOES })
      ],
      warning: NO_UNDO,
      confirmLabel: CONFIRM,
      confirmIcon: 'trash',
      destructive: true
    },
    handlers
  );
}
